File: problems_051to075/problem_071/solution.py
import os
import sys
from math import gcd

PROJECT_PATH = os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
sys.path.append(PROJECT_PATH)

from measurement import Stopwatch


class Solution(Stopwatch):
    def __init__(self):
        super().__init__()

    def execute(self):
        answer = None

        max_ratio = 0

        for denominator in range(2, 1000000+1):
            max_numerator = denominator * 3 // 7

            if max_numerator == denominator * 3 / 7:
                max_numerator -= 1

            for numerator in range(max_numerator, 0, -1):
                if gcd(numerator, denominator) == 1:
                    ratio = numerator / denominator

                    if max_ratio < ratio:
                        max_ratio = ratio
                        answer = numerator

                    break

        return answer

if __name__ == '__main__':
    soln = Solution()
    result = soln.execute()
    print(result)

File: problems_051to075/problem_071/solution.ts
import { SolutionFunction } from '../../lib/types';
import { gcd } from '../../lib/toolbox';

const solution: SolutionFunction = () => {
  let answer: ReturnType<SolutionFunction> = null;

  let maxRatio = 0;

  for (let denominator = 2; denominator <= 1000000; denominator += 1) {
    let maxNumerator = Math.floor((denominator * 3) / 7);

    if (maxNumerator === (denominator * 3) / 7) {
      maxNumerator -= 1;
    }

    for (let numerator = maxNumerator; numerator > 0; numerator -= 1) {
      if (gcd(numerator, denominator) === 1) {
        const ratio = numerator / denominator;

        if (maxRatio < ratio) {
          maxRatio = ratio;
          answer = numerator;
        }

        break;
      }
    }
  }

  return answer;
};

export default solution;
